pwiz_tools/Skyline/Model/Results/Scoring/AutoTrainManager.cs

```diff
@@ -0,0 +1,176 @@
+﻿/*
+ * Original author: Kaipo Tamura <kaipot .at. uw.edu>,
+ *                  MacCoss Lab, Department of Genome Sciences, UW
+ *
+ * Copyright 2021 University of Washington - Seattle, WA
+ * 
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using pwiz.Common.SystemUtil;
+using pwiz.Skyline.Model.DocSettings;
+using pwiz.Skyline.Model.DocSettings.Extensions;
+using pwiz.Skyline.Properties;
+using pwiz.Skyline.Util;
+using pwiz.Skyline.Util.Extensions;
+
+namespace pwiz.Skyline.Model.Results.Scoring
+{
+    public class AutoTrainManager : BackgroundLoader, IFeatureScoreProvider
+    {
+        private SrmDocument _document;
+        private IList<IPeakFeatureCalculator> _cacheCalculators;
+        private PeakTransitionGroupFeatureSet _cachedFeatureScores;
+
+        public override void ClearCache()
+        {
+        }
+
+        protected override bool StateChanged(SrmDocument document, SrmDocument previous)
+        {
+            return document.Settings.PeptideSettings.Integration.IsAutoTrain;
+        }
+
+        protected override string IsNotLoadedExplained(SrmDocument document)
+        {
+            if (document.Settings.PeptideSettings.Integration.IsAutoTrain &&
+                document.Settings.HasResults && document.MeasuredResults.IsLoaded)
+            {
+                return @"AutoTrainManager: Model not trained";
+            }
+            return null;
+        }
+
+        protected override IEnumerable<IPooledStream> GetOpenStreams(SrmDocument document)
+        {
+            yield break;
+        }
+
+        protected override bool IsCanceled(IDocumentContainer container, object tag)
+        {
+            return false;
+        }
+
+        protected override bool LoadBackground(IDocumentContainer container, SrmDocument document, SrmDocument docCurrent)
+        {
+            SrmDocument docNew;
+            var loadMonitor = new LoadMonitor(this, container, container.Document);
+
+            bool End(string message)
+            {
+                // Show an error message and set the AutoTrain flag to false.
+                IProgressStatus status = new ProgressStatus();
```

Comment:
Usually, you need to keep using the same ProgressStatus object all through the process, and pass it to everyone when you also pass the LoadMonitor.
Otherwise, the status bar ends up thinking that some part of the process has never ended.

I guess the way the code is here, each of the steps of this like "Score Features" and "Change Peaks" will appear as a separate task going from 0 to 100%, and nothing bad will happen, so I guess that's ok.